### PR TITLE
Add actions to control the new workload

### PR DESCRIPTION
**Pull request by nrobinaubertin**
### Overview

- Add actions to create/check/delete/list the ACLs
- Add a way to sideload a snap file for testing purposes. This does not work with the current libjuju without patching it, cf: https://github.com/juju/python-libjuju/pull/1147

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/bind.py [hunk 1]
```diff
@@ -104,16 +109,27 @@ def stop(self) -> None:
             logger.error(error_msg)
             raise StopError(error_msg) from e
 
-    def setup(self, unit_name: str) -> None:
+    def setup(self, unit_name: str, snap_path: str) -> None:
```
Comment (arturo-seijas) [inline src/bind.py:112]: Shouldn't this be optional?

File: src/charm.py [hunk 2]
```diff
@@ -66,6 +66,74 @@ def __init__(self, *args: typing.Any):
         self.unit.open_port("tcp", 8080)  # ACL API
         self.unit.open_port("tcp", 53)  # Bind DNS
         self.unit.open_port("udp", 53)  # Bind DNS
+        self.framework.observe(self.on.create_acl_action, self._on_create_acl_action)
```
Comment (arturo-seijas) [inline src/charm.py:69]: I don't know if this is a hard requirement, but iirc we are putting the actions code into an `actions` module

File: src/charm.py [hunk 3]
```diff
@@ -143,11 +211,18 @@ def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
 
     def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
         """Handle changed configuration event."""
+        self.unit.status = ops.MaintenanceStatus("Configuring workload")
+        self.bind.configure(
+            {
+                "django-debug": "true" if self.config["django_debug"] else "false",
+                "django-allowed-hosts": self.config["django_allowed_hosts"],
```
Comment (arturo-seijas) [inline src/charm.py:218]: It might be worth to validate this setting

File: src/bind.py [hunk 4]
```diff
@@ -104,16 +109,27 @@ def stop(self) -> None:
...
-            snap_name=constants.DNS_SNAP_NAME,
-            snap_channel=constants.SNAP_PACKAGES[constants.DNS_SNAP_NAME]["channel"],
-        )
+        # If a snap resource was not given, install the snap as published on snapcraft
+        if snap_path == "":
+            self._install_snap_package(
+                snap_name=constants.DNS_SNAP_NAME,
+                snap_channel=constants.SNAP_PACKAGES[constants.DNS_SNAP_NAME]["channel"],
+            )
+        else:
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply src/bind.py:129]: Done

File: src/charm.py (hunk 2 above)
Comment (nrobinaubertin) [reply src/charm.py]: After thinking about it, I decided to use a mixin to regroup functionalities around actions. I may extract more logic from the main charm file with the same pattern going forward

File: config.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+options:
+  django_debug:
+    type: boolean
+    description: "Set the backed by Django API DEBUG setting"
+    default: false
+  django_allowed_hosts:
```
Comment (amandahla) [inline config.yaml:8]: Couldn't be a comma-separated list like Paas app charmer does?
https://github.com/canonical/paas-app-charmer/blob/main/examples/django/charm/charmcraft.yaml#L38

File: tests/integration/test_charm.py [hunk 6]
```diff
@@ -18,6 +20,88 @@
...
+    assert unit.workload_status == ops.model.ActiveStatus.name  # type: ignore
+
+    for _, unit_ip in enumerate(await tests.integration.helpers.get_unit_ips(ops_test, unit)):
+        url = f"http://{unit_ip}:8080/static/admin/css/base.css"
+        try:
+            response = requests.head(url, timeout=5)
+            response.raise_for_status()
+            logger.info("File found at %s", url)
+        except requests.exceptions.HTTPError as e:
+            logger.info("HTTP Error: %s", e)
```
Comment (amandahla) [inline tests/integration/test_charm.py:44]: Will the test fail in case of HTTP error?

File: src/actions_mixin.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            self.on.list_acl_action, self._on_list_acl_action  # type: ignore[arg-type]
+        )
+
+    def _on_create_acl_action(self, event: ops.charm.ActionEvent) -> None:
+        """Handle the create ACL ActionEvent.
+
+        Args:
+            event: Event triggering this action handler.
+        """
+        event.set_results(
```
Comment (amandahla) [inline src/actions_mixin.py:50]: What happens if the command fails?

File: tests/integration/test_scale.py [hunk 8]
```diff
@@ -4,8 +4,8 @@
 
 """Integration tests."""
 
-# Ignore functions having too many arguments (will be removed in the future)
-# pylint: disable=too-many-positional-arguments
+# Ignore too many args warning. I NEED THEM OKAY ?!
```
Comment (amandahla) [inline tests/integration/test_scale.py:7]: Hahaha ;-)

**Comment by amandahla** [comment]
Nice PR!

Some points:
- I missed unit tests and failure tests
- For this one or a further PR: documentation on how to use charmed-bind-snap

File: tests/integration/test_charm.py (hunk 6 above)
Comment (nrobinaubertin) [reply tests/integration/test_charm.py]: You're right: it would not fail. I've removed the try/except block

File: src/actions_mixin.py (hunk 7 above)
Comment (nrobinaubertin) [reply src/actions_mixin.py:50]: I handled possible failures of `bind.command` so that a result is always returned

File: config.yaml (hunk 5 above)
Comment (nrobinaubertin) [reply config.yaml:8]: I changed that to mimic paas app charmer and be a comma-separated list

**Comment by nrobinaubertin** [comment]
>     * I missed unit tests and failure tests
There's not a a lot of added logic in the charm for now, everything is mostly handled by the snap who has its own tests.  
New tests will come when actually using these ACLs in the next PR

>     * For this one or a further PR: documentation on how to use charmed-bind-snap
This will be done in the next cycle for sure

**Review by amandahla — DISMISSED** [review]
LGTM

File: src/charm.py (hunk 3 above)
Comment (nrobinaubertin) [reply src/charm.py]: I added a test, It needs https://github.com/canonical/charmed-bind/pull/21 to fully work

**Review by amandahla — DISMISSED** [review]

**Review by javierdelapuente — COMMENTED** [review]
A few comments

File: src/charm.py [hunk 9]
```diff
@@ -66,6 +68,21 @@ def __init__(self, *args: typing.Any):
...
+        actions_mixin.ActionsMixin.hooks(self)
+
+        # Try to check if the `charmed-bind-snap` resource is defined.
+        # Using this can be useful when debugging locally
+        # More information about resources:
+        # https://juju.is/docs/sdk/resources#heading--add-a-resource-to-a-charm
+        self.snap_path: str = ""
+        try:
+            self.snap_path = str(self.model.resources.fetch("charmed-bind-snap"))
+        except ops.ModelError as e:
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply src/charm.py:80]: Yes, the CI is working without it. I'm just using it locally

File: tests/integration/test_charm.py [hunk 10]
```diff
@@ -18,6 +20,105 @@
...
 
+@pytest.mark.asyncio
+@pytest.mark.abort_on_fail
+async def test_admin_reachable(app: ops.model.Application, ops_test: OpsTest):
+    """
+    arrange: build and deploy the charm.
+    act: nothing.
+    assert: that the admin of the API is reachable.
+    """
+    # Application actually does have units
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply tests/integration/test_charm.py]: That explains a lot !

File: tests/integration/test_charm.py [hunk 11]
```diff
@@ -18,6 +20,105 @@
...
+    await app.set_config({"django_allowed_hosts": ""})  # type: ignore
+    time.sleep(10)
+
+    # Test that the admin is not reachable
+    for _, unit_ip in enumerate(await tests.integration.helpers.get_unit_ips(ops_test, unit)):
+        url = f"http://{unit_ip}:8080/admin/"
+        try:
+            response = requests.head(url, timeout=5)
+            response.raise_for_status()
+            assert False, f"Asset unexpectedly accessible at {url} after config change"
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply tests/integration/test_charm.py]: I forgot that it existed

File: src/charm.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply src/charm.py:80]: Okay, my bad. Doesn't work in the last CI anymore. Fixing...

File: src/charm.py [hunk 12]
```diff
@@ -31,7 +33,7 @@ class PeerRelationNetworkUnavailableError(exceptions.BindCharmError):
     """Exception raised when the peer relation network is unavailable."""
 
 
-class BindCharm(ops.CharmBase):
+class BindCharm(actions_mixin.ActionsMixin, ops.CharmBase):
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply src/charm.py:36]: IMO, if we pass and instance of the class, they are tightly coupled. I'm doing the mixin dance to acknowledge it but still split the code in two.  
This is in contrast to the bind service who can be tested outside of the charm instance.

**Comment by nrobinaubertin** [comment]
Closing this as ACLs will be part of a DNS policy charm

**Pull request closed without merging**